src/Marten/Events/Daemon/ProjectionDaemon.cs

```diff
@@ -56,549 +56,312 @@ await using (session.ConfigureAwait(false))
+
+            var request = new SubscriptionExecutionRequest(0, ShardExecutionMode.Rebuild, errorOptions, this);
+            await agent.ReplayAsync(request, highWaterMark, shardTimeout).ConfigureAwait(false);
+
+            _agents = _agents.AddOrUpdate(agent.Name.Identity, agent);
+        }
+        finally
         {
-            await StartDaemon().ConfigureAwait(false);
+            _semaphore.Release();
         }
-
-        var shards = _store.Options.Projections.AllShards();
-        foreach (var shard in shards)
-            await StartShard(shard, ShardExecutionMode.Continuous, CancellationToken.None).ConfigureAwait(false);
     }
 
     public async Task StartShard(string shardName, CancellationToken token)
     {
         if (!_highWater.IsRunning)
         {
-            await StartDaemon().ConfigureAwait(false);
+            await StartHighWaterDetectionAsync().ConfigureAwait(false);
         }
 
-        // Latch it so it doesn't double start
-        if (_agents.ContainsKey(shardName))
-        {
-            return;
-        }
+        var agent = _factory.BuildAgentForShard(shardName, Database);
+        var didStart = await tryStartAgentAsync(agent, ShardExecutionMode.Continuous).ConfigureAwait(false);
 
-        if (_store.Options.Projections.TryFindAsyncShard(shardName, out var shard))
+        if (!didStart && agent is IAsyncDisposable d)
         {
-            await StartShard(shard, ShardExecutionMode.Continuous, token).ConfigureAwait(false);
+            // Could not be started
+            await d.DisposeAsync().ConfigureAwait(false);
         }
     }
 
-    public async Task StopShard(string shardName, Exception? ex = null)
+    private async Task stopIfRunningAsync(string shardIdentity)
     {
-        if (_agents.TryGetValue(shardName, out var agent))
+        if (_agents.TryFind(shardIdentity, out var agent))
         {
-            if (agent.IsStopping)
+            var cancellation = new CancellationTokenSource();
+            cancellation.CancelAfter(5.Seconds());
+
+            try
             {
-                return;
+                await agent.StopAndDrainAsync(cancellation.Token).ConfigureAwait(true);
+            }
+            catch (Exception e)
+            {
+                Logger.LogError(e, "Error trying to stop and drain a subscription agent for '{Name}'",
+                    agent.Name.Identity);
             }
+            finally
+            {
+                _agents = _agents.Remove(shardIdentity);
+            }
+        }
+    }
 
-            var parameters = new ActionParameters(agent, async () =>
+    public async Task StopAsync(string shardName, Exception ex = null)
+    {
+        if (_agents.TryFind(shardName, out var agent))
+        {
+            await _semaphore.WaitAsync(_cancellation.Token).ConfigureAwait(false);
+            try
             {
+                var cancellation = new CancellationTokenSource();
```

Comment:
Should this be a [linked cancellation token source](https://learn.microsoft.com/en-us/dotnet/api/system.threading.cancellationtokensource.createlinkedtokensource?view=net-8.0)?
I.e. if cancellation happens externally (via the passed in token) or through the timeout (5 seconds) both will cancel the action.

The same on other places here.

Reply:
Yeah, I think there was a comment about that in at least one place. I'll make this one today